Sources/Profile-View.php

```diff
@@ -217,221 +217,362 @@ function summary($memID)
+		// Next, build the message strings.
+		foreach ($formats as $msg_type => $format_info)
+		{
+			// Get the values to use in the formatted string, in the right order.
+			$msg_values = array_replace(
+				array_fill_keys($format_info['required'], ''),
+				array_intersect_key($alert['extra'], array_flip($format_info['required']))
+			);
 
-		if (!empty($memberContext[$alert['sender_id']]))
-			$alerts[$id_alert]['sender'] = &$memberContext[$alert['sender_id']];
+			// Assuming all required values are present, build the message.
+			if (!in_array('', $msg_values))
+				$alert['extra'][$msg_type] = vsprintf($formats[$msg_type][$alert['show_links'] ? 'link' : 'text'], $msg_values);
 
+			elseif (in_array($msg_type, array('msg_msg', 'topic_msg', 'board_msg')))
+				$alert['extra'][$msg_type] = $txt[$msg_type == 'board_msg' ? 'board_na' : 'topic_na'];
+			else
+				$alert['extra'][$msg_type] = '(' . $txt['not_applicable'] . ')';
+		}
+
+		// Finally, set this alert's text string.
 		$string = 'alert_' . $alert['content_type'] . '_' . $alert['content_action'];
+
+		// This kludge exists because the alert content_types prior to 2.1 RC3 were a bit haphazard.
+		// This can be removed once all the translated language files have been updated.
+		if (!isset($txt[$string]))
+		{
+			if (strpos($alert['content_action'], 'unapproved_') === 0)
+				$string = 'alert_' . $alert['content_action'];
+
+			if ($alert['content_type'] === 'member' && in_array($alert['content_action'], array('report', 'report_reply')))
+				$string = 'alert_profile_' . $alert['content_action'];
+
+			if ($alert['content_type'] === 'member' && $alert['content_action'] === 'buddy_request')
+				$string = 'alert_buddy_' . $alert['content_action'];
+		}
+
 		if (isset($txt[$string]))
 		{
-			$extra = $alerts[$id_alert]['extra'];
-			$search = array('{member_link}', '{scripturl}');
-			$repl = array(!empty($alert['sender_id']) && $alert['show_links'] ? '<a href="' . $scripturl . '?action=profile;u=' . $alert['sender_id'] . '">' . $alert['sender_name'] . '</a>' : '<strong>'.$alert['sender_name'].'</strong>', $scripturl);
+			$substitutions = array(
+				'{scripturl}' => $scripturl,
+				'{member_link}' => !empty($sender_id) && $alert['show_links'] ? '<a href="' . $scripturl . '?action=profile;u=' . $sender_id . '">' . $alert['sender_name'] . '</a>' : '<strong>' . $alert['sender_name'] . '</strong>',
+			);
 
-			if (is_array($extra))
+			if (is_array($alert['extra']))
 			{
-				foreach ($extra as $k => $v)
-				{
-					$search[] = '{' . $k . '}';
-					$repl[] = $v;
-				}
+				foreach ($alert['extra'] as $k => $v)
+					$substitutions['{' . $k . '}'] = $v;
 			}
-			$alerts[$id_alert]['text'] = str_replace($search, $repl, $txt[$string]);
+
+			$alert['text'] = strtr($txt[$string], $substitutions);
 		}
+
+		// Unset the reference variable to avoid any surprises in subsequent loops.
+		unset($alert);
 	}
 
 	return $alerts;
```

Comment:
I'm thinking it would be nice to either move the hook before the return or add a new one that would maybe be called **integrate_fetch_alerts_after** there are some data that cannot be edited from the current hook like the icon

Reply:
I just moved the call to set_alert_icon() to a point before the call to the hook. I don't see anything else that a hook would need access to.

Reply:
yes I think that is enough indeed